### Data Quality Assesment
Data Dirty / Messy
1. uniqueness
2. validity
3. consistancy
4. completeness

### Extract Data

In [1]:
import pandas as pd  # lib untuk olah data
from sqlalchemy import create_engine # bikin connection antara code dengan db
from sqlalchemy import text

In [2]:
df = pd.read_csv('C:/Users/User/Documents/project/Binar/pandas_and_etl/dataset/mart_data_cut.csv')

### Data Quality Assesment

#### Uniqueness

In [4]:
# sql group by
# select id, count(*) as freq
# from table 
# group by id 
# having freq > 1

# show data yang duplicate
df[df.duplicated()] # munculkan df[dengan index true dari logic]

User_ID Product_ID Gender    Age  Occupation City_Category  \
136    1000026  P00042242      M  26-35         7.0             B   
137    1000026  P00059542      M  26-35         7.0             B   
10001  1000090  P00029242      M    55+        13.0             C   
10002  1001519  P00313242      M  18-25         4.0             C   
10003  1001519  P00085942      M  18-25         4.0             C   
10004  1001519  P00229642      M  18-25         4.0             C   
10005  1001519  P00313242      M  18-25         4.0             C   

       Stay_In_Current_City_Years  Marital_Status Product_Category  Purchase  
136                             2               1                5      5411  
137                             2               1                2      9734  
10001                           1               0                8      7786  
10002                           1               0                8      9803  
10003                           1               0                2     13244  
10004                           1               0                8      7868  
10005                           1               0                8      9803

In [6]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
10001     True
10002     True
10003     True
10004     True
10005     True
Length: 10006, dtype: bool

In [5]:
# show jumlah data yg duplicate
sum(df.duplicated())

7

In [9]:
df[df.duplicated(subset = ['User_ID', 'Product_ID'])]

User_ID Product_ID Gender    Age  Occupation City_Category  \
136    1000026  P00042242      M  26-35         7.0             B   
137    1000026  P00059542      M  26-35         7.0             B   
10001  1000090  P00029242      M    55+        13.0             C   
10002  1001519  P00313242      M  18-25         4.0             C   
10003  1001519  P00085942      M  18-25         4.0             C   
10004  1001519  P00229642      M  18-25         4.0             C   
10005  1001519  P00313242      M  18-25         4.0             C   

       Stay_In_Current_City_Years  Marital_Status Product_Category  Purchase  
136                             2               1                5      5411  
137                             2               1                2      9734  
10001                           1               0                8      7786  
10002                           1               0                8      9803  
10003                           1               0                2     13244  
10004                           1               0                8      7868  
10005                           1               0                8      9803

#### Completeness

In [ ]:
# cek completeness 
# sql
# select nama_col, count(*)
# from table
# where nama_col is null
# group by nama_col

# pandas
# df.info() -> helicopter view

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10006 entries, 0 to 10005
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   User_ID                     10006 non-null  int64  
 1   Product_ID                  10006 non-null  object 
 2   Gender                      10006 non-null  object 
 3   Age                         10006 non-null  object 
 4   Occupation                  9997 non-null   float64
 5   City_Category               10006 non-null  object 
 6   Stay_In_Current_City_Years  10006 non-null  int64  
 7   Marital_Status              10006 non-null  int64  
 8   Product_Category            10006 non-null  object 
 9   Purchase                    10006 non-null  int64  
dtypes: float64(1), int64(4), object(5)
memory usage: 781.8+ KB


In [13]:
# checking data with null value
# df[df.isna().any()]

df[df.isna().any(axis = 1)]

User_ID Product_ID Gender    Age  Occupation City_Category  \
1338  1000216  P00129342      M  46-50         NaN             B   
1369  1000220  P00334242      M  26-35         NaN             A   
5898  1000956  P00134142      M  36-45         NaN             A   
5899  1000956  P00035542      M  36-45         NaN             A   
5900  1000956  P00176242      M  36-45         NaN             A   
5901  1000957  P00365242      M  36-45         NaN             C   
9856  1001500  P00271142      M  26-35         NaN             A   
9857  1001501  P00030942      M  26-35         NaN             A   
9858  1001501   P0097142      M  26-35         NaN             A   

      Stay_In_Current_City_Years  Marital_Status Product_Category  Purchase  
1338                           1               0                1     15714  
1369                           3               0                1     15828  
5898                           3               1                5      6921  
5899                           3               1                3      8288  
5900                           3               1                8      9907  
5901                           2               1                5      8761  
9856                           2               0                5      8864  
9857                           3               0                4      2130  
9858                           3               0               12      1761

In [14]:
# menghitung jumlah row dari df
# len(df[logic index])
len(df[df.isna().any(axis = 1)])

9

#### Consistancy

In [15]:
# show all columns names
df.columns

Index(['User_ID', 'Product_ID', 'Gender', 'Age', 'Occupation', 'City_Category',
       'Stay_In_Current_City_Years', 'Marital_Status', 'Product_Category',
       'Purchase'],
      dtype='object')

In [16]:
# unique value di tiap column
# df['nama_kol'].unique()

df['User_ID'].unique()

array([1000001, 1000002, 1000003, ..., 1001527, 1001528, 1001530],
      dtype=int64)

In [18]:
list(df['Product_ID'].unique())

['P00069042',
 'P00248942',
 'P00087842',
 'P00085442',
 'P00285442',
 'P00193542',
 'P00184942',
 'P00346142',
 'P0097242',
 'P00274942',
 'P00251242',
 'P00014542',
 'P00031342',
 'P00145042',
 'P00231342',
 'P00190242',
 'P0096642',
 'P00058442',
 'P00036842',
 'P00249542',
 'P00220442',
 'P00156442',
 'P00213742',
 'P00214442',
 'P00303442',
 'P00135742',
 'P00039942',
 'P00161442',
 'P00078742',
 'P00085942',
 'P00118742',
 'P00297942',
 'P00266842',
 'P00058342',
 'P00032442',
 'P00105942',
 'P00182642',
 'P00186942',
 'P00155442',
 'P00221342',
 'P00087242',
 'P00111142',
 'P00259342',
 'P0094542',
 'P00148642',
 'P00312142',
 'P00113242',
 'P00192642',
 'P00110842',
 'P00189642',
 'P00304242',
 'P00365242',
 'P00129542',
 'P00140742',
 'P00182342',
 'P00276642',
 'P00334242',
 'P00247542',
 'P00338442',
 'P00275142',
 'P00333042',
 'P00166242',
 'P00161942',
 'P00348242',
 'P00042142',
 'P00244242',
 'P00217742',
 'P00019342',
 'P00100642',
 'P00073842',
 'P00366542',
 'P001907

In [19]:
list(df['Gender'].unique())

['F', 'M', 'Male', 'Female']

In [20]:
df[['User_ID','Gender']].groupby(by=['Gender']).nunique()

User_ID
Gender         
F           369
Female        1
M           992
Male          2

In [21]:
list(df['Age'].unique())

['0-17', '55+', '26-35', '46-50', '51-55', '36-45', '18-25']

#### Validity

In [27]:
df.describe([.25,.5,.75, .98]).apply(lambda s: s.apply('{0:.1f}'.format))

User_ID Occupation Stay_In_Current_City_Years Marital_Status  \
count    10006.0     9997.0                    10006.0        10006.0   
mean   1000791.1        8.4                        1.9            0.4   
std        440.5        6.7                        1.3            0.5   
min    1000001.0        0.0                        0.0            0.0   
25%    1000403.0        3.0                        1.0            0.0   
50%    1000817.0        7.0                        2.0            0.0   
75%    1001172.0       15.0                        3.0            1.0   
98%    1001495.0       20.0                        4.0            1.0   
max    1001530.0       20.0                        4.0            1.0   

         Purchase  
count     10006.0  
mean      17737.1  
std      783231.7  
min         186.0  
25%        5832.0  
50%        8022.0  
75%       11923.0  
98%       19959.5  
max    77950291.0

In [28]:
# df[logic]
df[df['Purchase'] > 100000]

User_ID Product_ID Gender    Age  Occupation City_Category  \
155   1000029  P00127242      M  36-45         7.0             C   
2898  1000466  P00031742      M  26-35         5.0             A   

      Stay_In_Current_City_Years  Marital_Status Product_Category  Purchase  
155                            1               0                1   7963012  
2898                           2               0                8  77950291

Total data 10006

Summary Quality assesment
1. ada 7 data yang duplicate
2. ada 9 data yang null, dan ini terjadi di kolom Occupation
3. nama kolom masih menggunakan capital, padahal snake case
4. pada kolom gender ada 2 data yang menggunakan value 'Male' dan 1 data 'Female', sisanya menggunakan F dan M
5. dari validity terdapat 2 data yang purchase > 7 jt usd

### Data Clensing

#### Remove duplication

In [30]:
df_clean = df.copy()

In [31]:
df_clean = df_clean.drop_duplicates(keep = 'last')

In [36]:
df_clean[df_clean.duplicated()]

Empty DataFrame
Columns: [User_ID, Product_ID, Gender, Age, Occupation, City_Category, Stay_In_Current_City_Years, Marital_Status, Product_Category, Purchase]
Index: []

#### remove null row

In [37]:
df_clean = df_clean.dropna()

In [38]:
df_clean[df_clean.isna().any(axis = 1)]

Empty DataFrame
Columns: [User_ID, Product_ID, Gender, Age, Occupation, City_Category, Stay_In_Current_City_Years, Marital_Status, Product_Category, Purchase]
Index: []

### Replace unconsistance data

In [41]:
# df['namacolumn'].replace({'yang_mau_direplace1' : 'hasil_replace1'})
df_clean['Gender'] = df_clean['Gender'].replace({'Male':'M', 'Female':'F'})

In [42]:
list(df_clean['Gender'].unique())

['F', 'M']

#### Additional matery

In [43]:
pdf = pd.read_csv('C:/Users/User/Documents/project/Binar/pandas_and_etl/dataset/sales_market/sales_order.csv') 

In [65]:
# change from string to date
pdf['constructed_date'] = pd.to_datetime(pdf['Date'], format='%m/%d/%Y').dt.date

In [66]:
pdf['constructed_date'].min()

datetime.date(2001, 1, 1)

In [67]:
pdf['constructed_date'].max()

datetime.date(2019, 3, 30)

In [68]:
pdf.head()

Invoice ID Branch Customer type  Gender            Product line  \
0  750-67-8428      A        Member  Female       Health and beauty   
1  226-31-3081      C        Normal  Female  Electronic accessories   
2  631-41-3108      A        Normal    Male      Home and lifestyle   
3  123-19-1176      A        Member    Male       Health and beauty   
4  373-73-7910      A        Normal    Male       Sports and travel   

   Unit price  Quantity   Tax 5%        Date   Time      Payment    cogs  \
0       74.69         7  26.1415  01/05/2019  13:08      Ewallet  522.83   
1       15.28         5   3.8200  03/08/2019  10:29         Cash   76.40   
2       46.33         7  16.2155  03/03/2019  13:23  Credit card  324.31   
3       58.22         8  23.2880   1/27/2019  20:33      Ewallet  465.76   
4       86.31         7  30.2085  02/08/2019  10:37      Ewallet  604.17   

   gross income  Rating  const_date constructed_date  
0       26.1415     9.1  2019-01-05       2019-01-05  
1        3.8200     9.6  2019-03-08       2019-03-08  
2       16.2155     7.4  2019-03-03       2019-03-03  
3       23.2880     8.4  2019-01-27       2019-01-27  
4       30.2085     5.3  2019-02-08       2019-02-08